### PR TITLE
Child process

### PR DESCRIPTION
**Pull request by readikus**

### CODE REVIEW

File: child-processes/index.md [hunk 1]
```diff
@@ -1,8 +1,189 @@
 ---
-layout: default
+layout: post
 title: Child Processes
+url: child-processes
+author: ian
 ---
 
-Coming soon.
+<<<<<<< HEAD
```
Comment (donovanh) [inline child-processes/index.md]: Some merge artefact

File: child-processes/fork-child.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+const fibonacci = (num) => num <= 1 ? 1 : fibonacci(num - 1) + fibonacci(num - 2)
+
+process.on('message', ({n}) => {
+  process.send({ fib: fibonacci(n), n })
+  // optional - there is no reason why this child process
+  // can't be called multiple times.
+  process.exit()
+})
```
Comment (donovanh) [inline child-processes/fork-child.js:8]: Should this code have semi-colons?

File: child-processes/fork-parent.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+const { fork } = require('child_process')
+
+const child1 = fork('fork-child')
+const child2 = fork('fork-child')
+const child3 = fork('fork-child')
```
Comment (donovanh) [inline child-processes/fork-parent.js]: Semi-colons missing but present elsewhere

File: child-processes/index.md [hunk 4]
```diff
@@ -1,8 +1,189 @@
...
+| spawns a shell in which the command is executed            | Does not spawn a shell                                      |
+| buffers the data (waits until the process finishes and transfers all the data ) | Streams the data returned by the child process (data flow is constant) |
+| maximum data transfer 200kb (by default                    | has no data transfer size limit                             |
+
+Typically, `spawn` is more suitable for long-running process with large outputs. spawn streams input/output with child process. `exec` buffered output in a small (by default 200K) buffer.
+
+## A Simple `exec`
+
+In this example, we will call the `ls` command to listwith the optional param `-l` to show a long list of details. 
+<div class="repl-code">
```
Comment (donovanh) [inline child-processes/index.md:36]: Tiny nitpick: I'd adda line before the HTML code here to make it easier to scan

File: child-processes/index.md [hunk 5]
```diff
@@ -1,8 +1,189 @@
...
+|------------------------------------------------------------|-------------------------------------------------------------|
+| spawns a shell in which the command is executed            | Does not spawn a shell                                      |
+| buffers the data (waits until the process finishes and transfers all the data ) | Streams the data returned by the child process (data flow is constant) |
+| maximum data transfer 200kb (by default                    | has no data transfer size limit                             |
+
+Typically, `spawn` is more suitable for long-running process with large outputs. spawn streams input/output with child process. `exec` buffered output in a small (by default 200K) buffer.
+
+## A Simple `exec`
+
+In this example, we will call the `ls` command to listwith the optional param `-l` to show a long list of details.
```
Comment (donovanh) [inline child-processes/index.md]: Is `listwith` a thing? :D

File: child-processes/index.md [hunk 6]
```diff
@@ -1,8 +1,189 @@
...
+
+const { exec } = require('child_process');
+
+exec('ls -l', (err, stdout, stderr) => {
+  console.log(stdout);
+});
+
+```
+</div>
+
```
Comment (donovanh) [inline child-processes/index.md:48]: I'd maybe add a quick line saying what exec does here - may seem obvious but I think it helps to reinforce the point.

File: child-processes/index.md [hunk 7]
```diff
@@ -1,8 +1,189 @@
...
+});
+
+ls.on('close', (code) => {
+  console.log(`child process exited with code ${code}`);
+});
+
+```
+</div>
+
+In this example we are listening to the `stout` and `stderr` streams for `data` events, as well as listening for a `close` event.
```
Comment (donovanh) [inline child-processes/index.md]: Are these events different from the exec approach?

File: child-processes/index.md [hunk 8]
```diff
@@ -1,8 +1,189 @@
...
+wc.on('close', (code) => {
+  console.log(`wc child process wc exited with code ${code}`);
+});
+
+```
+</div>
+
+# fork
+
+`fork` is a special version of `spawn' that allows messages to be sent between the Node processes.
```
Comment (donovanh) [inline child-processes/index.md]: `spawn' -> `spawn`

File: child-processes/index.md [hunk 9]
```diff
@@ -1,8 +1,189 @@
...
+
+## A note on options
+
+These processes all accept an optional options object that allow us to control the context that the processes are run within. These vary for each method, and are described in detail in the [node documentation for `child_process`](https://nodejs.org/api/child_process.html).
+
+## Summary
+
+Parents `spawm`, `fork` or `exec` child processes, and communicate via events or pipes.
+
+Take me to [cheat sheet]({{ "/cheatsheet/#child-process" | url }})
```
Comment (donovanh) [inline child-processes/index.md]: I'd stick a full stop after the link here :)

File: child-processes/index.md [hunk 10]
```diff
@@ -1,8 +1,189 @@
...
+
+## Exercise
+
+Create a child process for doing some manipulation of a file or URL, and build a parent process that controls a number of these processes in parallel.
```
Comment (donovanh) [inline child-processes/index.md:194]: Should this have a hidden suggested-solution? If too much to do, we can leave it for now.

**Review by donovanh — CHANGES_REQUESTED** [review]
Looking good - I've just thrown in some nitpicks

File: child-processes/index.md (hunk 10 above)
Comment (readikus) [reply child-processes/index.md:194]: Yeah - I think there is more value in populating the other bits of content and we can come back to this, or maybe I can put a note in for someone to do one and put it up as a PR :)

File: child-processes/index.md (hunk 1 above)
Comment (readikus) [reply child-processes/index.md]: Sorted

File: child-processes/fork-child.js (hunk 2 above)
Comment (readikus) [reply child-processes/fork-child.js:8]: Sorted!

File: child-processes/index.md (hunk 4 above)
Comment (readikus) [reply child-processes/index.md:36]: Done

File: child-processes/index.md (hunk 6 above)
Comment (readikus) [reply child-processes/index.md:48]: Just pushed

File: child-processes/index.md (hunk 7 above)
Comment (readikus) [reply child-processes/index.md]: Yes - exec buffers it to strints

Comment (readikus) [reply child-processes/index.md]: Strings.

I'll add that :)

File: child-processes/index.md (hunk 10 above)
Comment (readikus) [reply child-processes/index.md:194]: Added comment - thanks for all the comments and keeping me honest. Made a big improvement to this piece.

**Review by donovanh — APPROVED** [review]
Let's give this a whirl. Good job

**Pull request merged**